Add `ScriptBuilder` to streamline the process of building note & tx scripts

partylikeits1983 opened the pull request:
### Feature description

#### **Background / Motivation**

Creating note and transaction (tx) scripts currently requires a fair amount of boilerplate.
This PR continues the refactor started in [miden-client #944](https://github.com/0xMiden/miden-client/pull/944) and replaces the older, now-stale PR in miden-base (#1467). The new approach introduces **`ScriptBuilder`**, a helper that hides assembler setup and library wiring so that developers can focus on the script itself. The motivation of creating the `ScriptBuilder` comes from @bobbinth 's  comment here: https://github.com/0xMiden/miden-client/pull/944#discussion_r2141304023

Resolves: #1506 

#### **What this PR does**

* Adds a `ScriptBuilder` implementation that
  * accepts an optional pre-configured `Assembler`
  * supports debug / non-debug mode toggling
  * manages an arbitrary number of libraries via `add_library`
  * exposes one-liners to compile:
    * `compile_library`
    * `compile_note_script`
    * `compile_tx_script`
  
* Includes unit tests that cover:
  * constructor variants
  * library compilation and attachment
  * basic script compilation flows

#### **Usage example**

```rust
use miden_objects::transaction::ScriptBuilder;

let mut builder = ScriptBuilder::new(true);        // debug mode
let lib = builder
    .compile_library(r#"export.inc begin push.1 add end"#, "math::lib")?;
builder.add_library(vec![lib])?;

let note_script = builder.compile_note_script("use.math::lib begin call.lib::inc end")?;
let tx_script   = builder.compile_tx_script("begin nop end")?;
```

For scripts that call kernel procedures, build the builder with a kernel assembler:

```rust
let kernel_asm = TransactionKernel::assembler();
let mut builder = ScriptBuilder::with_assembler(kernel_asm, false);
```

#### **PR relationships**

* **Unblocks** [miden-client #944](https://github.com/0xMiden/miden-client/pull/944)
* **Supersedes / closes** miden-base #1467


### Why is this feature needed?

The `ScriptBuilder` reduces the overhead when building notes & tx scripts.

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+    ///
+    /// # Arguments
+    /// * `in_debug_mode` - Whether to enable debug mode in the assembler
+    pub fn new(in_debug_mode: bool) -> Self {
+        let assembler = Assembler::default().with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
```
Given that scripts are very likely to use some functionality from the miden protocol libraries, I would create the assembler here as:
```rust
let assembler = TransactionKernel::assembler().with_debug_mode(in_debug_mode);
```

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    /// Creates a new ScriptBuilder with a provided assembler.
+    ///
+    /// This is the recommended constructor when you need access to transaction kernel
+    /// procedures. Pass `TransactionKernel::assembler()` as the assembler parameter.
+    ///
+    /// # Arguments
+    /// * `assembler` - A pre-configured assembler (e.g., from TransactionKernel::assembler())
+    /// * `in_debug_mode` - Whether to enable debug mode
+    pub fn with_assembler(assembler: Assembler, in_debug_mode: bool) -> Self {
+        let assembler = assembler.with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
```
Related to the previous comment - I'm not sure we need this constructor. I would rather completely abstract away the assembler via this struct.

partylikeits1983 replied:
Agreed. Will remove once we figure out the best way to handle this: https://github.com/0xMiden/miden-base/pull/1507#discussion_r2178545010

After second thought maybe the best way is to move the `ScriptBuilder` to `miden-lib`? 

@Fumuran @PhilippGackstatter Would love to get your feedback on this

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+/// For scripts that need access to transaction kernel procedures,
+/// use `ScriptBuilder::with_assembler()` and provide an assembler created with
+/// `TransactionKernel::assembler()` from the miden-lib crate.
+pub struct ScriptBuilder {
+    assembler: Assembler,
+    libraries: Vec<Library>,
+}
```
I would add `impl Default for ScriptBuilder` which would basically call `ScriptBuilder::new(false)`.

partylikeits1983 replied:
Added a default for the `ScriptBuilder`

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
nit: I'd add `TESTS` section separator above this line.

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+    /// - The transaction script compilation fails
+    pub fn compile_tx_script(
+        &self,
+        tx_script: &str,
```
Is there a reason not to use `impl AsRef<str>` instead of `&str` here? I think it would cover all the use cases as `&str` does but would be a bit more flexible.

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+    /// Returns an error if:
+    /// - Any of the libraries cannot be loaded into the assembler
+    /// - The note script compilation fails
+    pub fn compile_note_script(&self, program: &str) -> Result<NoteScript, ScriptBuilderError> {
```
Similar to the above comment: could we use `impl AsRef<str>` instead of `&str` here?

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+    /// - The library path is invalid
+    /// - The library code cannot be parsed
+    /// - The library cannot be assembled
+    pub fn compile_library(
+        &self,
+        library_code: &str,
+        library_path: &str,
+    ) -> Result<Library, ScriptBuilderError> {
```
Similar to the previous two comments, could we use the following signature:
```rust
pub fn compile_library(
    &self,
    library_code: impl AsRef<str>,
    library_path: impl TryInto<LibraryPath>,
) -> Result<Library, ScriptBuilderError>
```
Or does it create some complications?

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+        library_code: &str,
+        library_path: &str,
+    ) -> Result<Library, ScriptBuilderError> {
+        let assembler = self.assembler.clone().with_debug_mode(true);
```
Is there a reason to override the debug mode of the assembler here?

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,314 @@
+    ///
+    /// # Arguments
+    /// * `libraries` - Vector of compiled libraries to add to the builder
+    pub fn add_library(&mut self, libraries: Vec<Library>) -> Result<(), ScriptBuilderError> {
```
nit: instead of `Vec<Library>` we should be able to do something like `impl Iterator<Item = Library>` or maybe `impl IntoIterator<Item = Library>`.

bobbinth reviewed (commented):
Looks good! Thank you! I left some small comments inline.

PhilippGackstatter commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,331 @@
+        &mut self,
+        libraries: impl IntoIterator<Item = Library>,
+    ) -> Result<(), ScriptBuilderError> {
+        for library in libraries.into_iter() {
+            self.libraries.push(library);
+        }
```
Is there a reason why the libraries are not added to the assembler directly and kept in this intermediary `libraries` field? If this is to support instantiating a builder with some set of libraries once, and then cloning it to have a second builder with the same libraries, the same should work if the libraries are added directly to the assembler. That should avoid some extra copying around in memory.

partylikeits1983 replied:
Resolved!

PhilippGackstatter reviewed (commented):
Looks good!

I left a few comments, mostly related to the API and usage.

Primarily:
- We should be able to get away with a simple `struct` error type for this.
- We should point out that `miden` and `std` are _dynamically_ linked and all user-provided libraries are _statically_ linked. Ideally, that should be reflected in the API names.
- I'm not sure the intermediary `libraries` field is necessary, I think we can add those to the assembler directly.
- I think we should not allow compiling libraries through the script builder. That seems beyond the scope of a script builder and in order to do this job justice, it would essentially have to reexpose most `Assembler` APIs. I think `TransactionKernel::assembler` should be used for that purpose directly.

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,331 @@
+    ///
+    /// # Arguments
+    /// * `in_debug_mode` - Whether to enable debug mode in the assembler
+    pub fn new(in_debug_mode: bool) -> Self {
```
I would keep it for now as this mirrors what we have in the client. Once the assembler is updated, we can remove it.

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,331 @@
+    /// - The library path is invalid
+    /// - The library code cannot be parsed
+    /// - The library cannot be assembled
+    pub fn compile_library(
```
One of the motivations for introducing the script builder is to simplify building note/tx scripts in some common user-facing patterns. So, one of the goals is that the user should be able to do something like this without having to interact with the assembler (or transaction kernel) directly.

I think the a common pattern here is that the user may have some MASM code in string form that they want to use as a module/library in their note/tx script to keep things more modular. And ideally doing this shouldn't involve instantiating a new assembler with correct parameters (or even knowing about the assembler).

So, I think having a method like this is fine - but I do now think that maybe the naming should be a bit different. Specifically, what we are adding here is not a library but a module. So, maybe the method should look like:
```rust
pub fn link_module(
    &self,
    module_code: impl AsRef<str>,
    module_path: impl AsRef<str>,
) -> Result<(), ScriptBuilderError> 
```
This will add a statically-linked module to the assembler under the specified path to be available for use in note/tx scripts.

bobbinth commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,331 @@
+    ///
+    /// # Arguments
+    /// * `libraries` - Iterator of compiled libraries to add to the builder
+    pub fn add_library(
```
> We don't need an equivalent to `Assembler::link_dynamic_library` because the only dynamically linked libraries are the `miden` and `std` libraries, which should automatically be added to the assembler. User-written dynamically linked libraries would not be available on-chain, so it does not make sense to allow that. This would be worth pointing out in the docs to the builder as well.

Now that I think about it, maybe we do want the ability to link dynamic libraries as well. The main use case for this could be FPI calls. That is, if a note/tx script wants to make an FPI call, the code for this would be available on-chain, and so, we don't need to link it statically.

So, maybe indeed we should have two methods: `link_static_library()` and `link_dynamic_library()` (and we could have the corresponding `with_` methods for convenience).

PhilippGackstatter replied:
> The main use case for this could be FPI calls.

Oh yes, good point! That does make sense. Then we should clearly explain in the docs when to use static vs dynamic linking.

partylikeits1983 replied:
I implemented the `with_statically_linked_library` and `with_dynamically_linked_library` functions. Let me know if I can improve the implementation further.

partylikeits1983 commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,331 @@
+        &self,
+        tx_script: impl AsRef<str>,
+    ) -> Result<TransactionScript, ScriptBuilderError> {
+        let mut assembler = self.assembler.clone();
```
Removed the clone from the function, let me know if this approach is more efficient

partylikeits1983 commented:
@bobbinth @PhilippGackstatter 

I removed the `Vec<Library>` field on the `ScriptBuilder`. Also big thank you @PhilippGackstatter for outlining how the API for linking would look like, this made it easier to understand what needed to be added. I think I have addressed all of your suggestions, but take one more look to see if I missed anything.

partylikeits1983 commented on crates/miden-objects/src/transaction/script_builder.rs:
```diff
@@ -0,0 +1,331 @@
+    #[error("transaction script library error: {0}")]
+    TransactionScriptLibraryError(String),
+    #[error("transaction script error")]
+    TransactionScriptError(#[from] TransactionScriptError),
```
Makes sense. Modified as to not stringify the errors. Lmk if there is anything else I need to modify.

partylikeits1983 commented on crates/miden-lib/src/utils/script_builder.rs:
```diff
@@ -0,0 +1,306 @@
+
+        // Add the compiled library to our collection
+        self.libraries.push(library);
+        Ok(())
```
Fixed

igamigo commented on crates/miden-lib/src/errors/script_builder_errors.rs:
```diff
@@ -0,0 +1,58 @@
+
+impl ScriptBuilderError {
+    /// Creates a script builder error from an error message.
+    pub fn build_error(message: impl Into<String>) -> Self {
+        let message: String = message.into();
+        Self { message: message.into(), source: None }
+    }
```
This doesn't seem used, maybe it should be removed?

igamigo commented on crates/miden-lib/src/utils/script_builder.rs:
````diff
@@ -0,0 +1,458 @@
+///
+/// ## Builder Pattern Example
+///
+/// ```ignore
````
Here, it would be nice to use `no_run` instead of `ignore` so that the example is built (but not executed) if the API somehow changes, we can make sure the example is still compatible. You can hide some setup lines with `#` so that they get built but not rendered in the docs.

igamigo reviewed (commented):
Looking good! Leaving some minor comments for now, not too related to the actual builder functionality

partylikeits1983 commented on crates/miden-lib/src/utils/script_builder.rs:
```diff
@@ -0,0 +1,458 @@
+        self,
+        library_code: impl AsRef<str>,
+        library_path: impl AsRef<str>,
+    ) -> Result<Library, ScriptBuilderError> {
```
Yes, I realized it is redundant. Decided to remove it from the `ScriptBuilder`

partylikeits1983 commented on crates/miden-lib/src/utils/script_builder.rs:
```diff
@@ -0,0 +1,458 @@
+    ///
+    /// This creates a basic assembler. For transaction scripts that need access to
+    /// transaction kernel procedures, use `ScriptBuilder::with_assembler()` and provide
+    /// an assembler created with `TransactionKernel::assembler()`.
```
Edited the comment

partylikeits1983 commented on crates/miden-lib/src/utils/mod.rs:
```diff
@@ -0,0 +1,5 @@
+
+pub use script_builder::ScriptBuilder;
+
+pub use crate::errors::ScriptBuilderError;
```
This is much cleaner now.

bobbinth reviewed (approved):
All looks good! Thank you!

PhilippGackstatter reviewed (commented):
Looks good, thank you!

I really like the description of static vs dynamic linking and the builder-style example - feels very fluent.

Left some comments. I think the one about the tests and the exports of the transaction kernel and note script errors is important, the others are just nits.

partylikeits1983 commented on crates/miden-lib/src/lib.rs:
```diff
@@ -17,15 +17,13 @@ mod auth;
 pub use auth::AuthScheme;
 
 pub mod account;
-#[cfg(any(feature = "testing", test))]
```
Added `#[cfg(any(feature = "testing", test))]` to `tx_kernel_errors `and `note_script_errors` modules to keep auto-generated errors behind testing feature as suggested.

partylikeits1983 commented on crates/miden-lib/src/utils/script_builder.rs:
```diff
@@ -0,0 +1,420 @@
+        let library_path = "external_contract::counter_contract";
+
+        let mut builder_with_lib = ScriptBuilder::new(true);
+        let add_result = builder_with_lib.link_module(library_path, account_code);
+
+        if add_result.is_ok() {
+            let _tx_script_result = builder_with_lib.compile_tx_script(script_code);
+        }
```
Good point. Updated all test functions to return `anyhow::Result<()>` and use proper error handling with `.context()`

PhilippGackstatter reviewed (approved):
Looks great, thank you!
